File: src/com/ashcraftmedia/daftman/tile/Tile.java
package com.ashcraftmedia.daftman.tile;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Point;

public abstract class Tile {
	public static final Dimension size = new Dimension(32, 32);
	
	private int row;
	private int col;
	
	/**
	 * Constructor for tile objects. Decreases the row and column by
	 * one, so the board has a negative row and a negative column.
	 * 
	 * @param aRow The row of the tile
	 * @param aCol The col of the tile
	 */
	public Tile(int aRow, int aCol) {
		row = aRow-1;
		col = aCol-1;
	}
	
	/**
	 * Draws.
	 * 
	 * @param g The Graphics object
	 */
	public abstract void draw(Graphics g);
	
	public Dimension getSize() {
		return size;
	}
	
	/**
	 * Returns the origin location.
	 * 
	 * @return The origin location
	 */
	public Point getLoc() {
		return new Point(col * size.width, row * size.height);
	}
	
	/**
	 * Returns the center location.
	 * 
	 * @return The center location
	 */
	public Point getCenter() {
		return new Point(getLoc().x + size.width/2, getLoc().y + size.height/2);
	}
	
	/**
	 * Returns if the tile is impassable.
	 * 
	 * @return If the tile is impassable
	 */
	public boolean isImpassable() {
		return false;
	}
	
	/**
	 * Returns if the tile is destructible.
	 * 
	 * @return If the tile is destructible
	 */
	public boolean isDestructible() {
		return false;
	}

	public int getRow() {
		return row;
	}

	public int getCol() {
		return col;
	}
	
	public boolean isAdjacent(Tile t) {
		return Math.abs(row - t.getRow()) < 1 && Math.abs(col - t.getCol()) < 1;
	}
	
	public boolean equals(Object o) {
		if (o != null && o instanceof Tile) {
			Tile other = (Tile)o;
			return other.getCenter().equals(getCenter());
		}
		
		return false;
	}
	
	public String toString() {
		return "Tile @ ("+row+", "+col+")";
	}
}

File: web/js/tile/Tile.js
export default class Tile {
    static size = { width: 32, height: 32 };

    constructor(aRow, aCol) {
        this._row = aRow;
        this._col = aCol;
    }

    draw(ctx) {
        throw new Error('draw() must be implemented by subclass');
    }

    getSize() {
        return Tile.size;
    }

    getLoc() {
        return {
            x: this._col * Tile.size.width,
            y: this._row * Tile.size.height
        };
    }

    getCenter() {
        const loc = this.getLoc();
        return {
            x: loc.x + Tile.size.width / 2,
            y: loc.y + Tile.size.height / 2
        };
    }

    isImpassable() {
        return false;
    }

    isDestructible() {
        return false;
    }

    getRow() {
        return this._row;
    }

    getCol() {
        return this._col;
    }

    isAdjacent(t) {
        return Math.abs(this._row - t.getRow()) < 1 && Math.abs(this._col - t.getCol()) < 1;
    }

    equals(o) {
        if (o != null && o instanceof Tile) {
            const otherCenter = o.getCenter();
            const thisCenter = this.getCenter();
            return otherCenter.x === thisCenter.x && otherCenter.y === thisCenter.y;
        }
        return false;
    }

    toString() {
        return `Tile @ (${this._row}, ${this._col})`;
    }
}
